File: src/geometric_ops.py
import cv2
import numpy as np

class GeometricOps:
    @staticmethod
    def crop(frames, x, y, w, h):
        return [f[y:y+h, x:x+w] for f in frames]
    
    @staticmethod
    def resize(frames, w, h):
        return [cv2.resize(f, (w, h)) for f in frames]
    
    @staticmethod
    def rotate(frames, angle, w, h):
        m = cv2.getRotationMatrix2D((w//2, h//2), angle, 1)
        return [cv2.warpAffine(f, m, (w, h)) for f in frames]
    
    @staticmethod
    def flip(frames, direction):
        return [cv2.flip(f, direction) for f in frames]
    
    @staticmethod
    def zoom(frames, factor, center_x=None, center_y=None):
        result = []
        for f in frames:
            h, w = f.shape[:2]
            if center_x is None:
                center_x = w // 2
            if center_y is None:
                center_y = h // 2
            new_w = int(w / factor)
            new_h = int(h / factor)
            x1 = max(0, center_x - new_w // 2)
            y1 = max(0, center_y - new_h // 2)
            x2 = min(w, x1 + new_w)
            y2 = min(h, y1 + new_h)
            cropped = f[y1:y2, x1:x2]
            result.append(cv2.resize(cropped, (w, h)))
        return result
    
    @staticmethod
    def pan(frames, dx, dy):
        result = []
        for f in frames:
            h, w = f.shape[:2]
            M = np.float32([[1, 0, dx], [0, 1, dy]])
            result.append(cv2.warpAffine(f, M, (w, h)))
        return result

File: src/GeometricOps.cpp
#include "../include/GeometricOps.h"

using namespace cv;

std::vector<Mat> GeometricOps::crop(const std::vector<Mat>& frames, int x, int y, int w, int h) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Rect roi(x, y, w, h);
        result.push_back(f(roi).clone());
    }
    return result;
}

std::vector<Mat> GeometricOps::resize(const std::vector<Mat>& frames, int w, int h) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat resized;
        cv::resize(f, resized, Size(w, h));
        result.push_back(resized);
    }
    return result;
}

std::vector<Mat> GeometricOps::rotate(const std::vector<Mat>& frames, double angle, int w, int h) {
    std::vector<Mat> result;
    Point2f center(w / 2.0f, h / 2.0f);
    Mat M = getRotationMatrix2D(center, angle, 1.0);
    for (const auto& f : frames) {
        Mat rotated;
        warpAffine(f, rotated, M, Size(w, h));
        result.push_back(rotated);
    }
    return result;
}

std::vector<Mat> GeometricOps::flip(const std::vector<Mat>& frames, int direction) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        Mat flipped;
        cv::flip(f, flipped, direction);
        result.push_back(flipped);
    }
    return result;
}

std::vector<Mat> GeometricOps::zoom(const std::vector<Mat>& frames, double factor, int centerX, int centerY) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        int h = f.rows, w = f.cols;
        int cx = centerX == -1 ? w / 2 : centerX;
        int cy = centerY == -1 ? h / 2 : centerY;
        int newW = (int)(w / factor);
        int newH = (int)(h / factor);
        int x1 = std::max(0, cx - newW / 2);
        int y1 = std::max(0, cy - newH / 2);
        int x2 = std::min(w, x1 + newW);
        int y2 = std::min(h, y1 + newH);
        Rect roi(x1, y1, x2 - x1, y2 - y1);
        Mat cropped = f(roi);
        Mat resized;
        cv::resize(cropped, resized, Size(w, h));
        result.push_back(resized);
    }
    return result;
}

std::vector<Mat> GeometricOps::pan(const std::vector<Mat>& frames, int dx, int dy) {
    std::vector<Mat> result;
    for (const auto& f : frames) {
        int h = f.rows, w = f.cols;
        Mat M = (Mat_<float>(2,3) << 1, 0, dx, 0, 1, dy);
        Mat panned;
        warpAffine(f, panned, M, Size(w, h));
        result.push_back(panned);
    }
    return result;
}
